### ISSUE TITLE
Performance improvements for take

### ISSUE THREAD

**Issue by Dandandan** [apache/arrow-rs#279]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Take is a critical function that is used in quite some code and is hot up in profiling. It seems like we should be able to speed it up.

Two improvements we can do:
* Bound checks are about 25% of the work (in take benchmark)
* SIMD could be utilized.

**Describe the solution you'd like**
Check whether we can eliminate / avoid bound checks.
Add SIMD  versions of take 

**Describe alternatives you've considered**

**Additional context**

**Comment by Dandandan** [comment]
Some first experimenting didn't reveal a lot of options.
What did improve performance however, was removing the bound check (-25%) but it doesn't seem like we can get rid of it without checking bounds for all values manual (which seems like it should be slower - even after speeding it up by 2x here https://github.com/apache/arrow-rs/pull/281 )

**Comment by ritchie46** [comment]
> It seems further speedups might come only from somehow removing bound checks and SIMD gather.

SIMD gather is not being used atm right? I remember looking for this operation in `packed_simd`, but not yielding any results.

**Comment by Dandandan** [comment]
Yeah I also did some googling - seems not supported in `packed_simd` so I guess at this point it has to be written by hand.

**Comment by nevi-me** [comment]
These are good opportunities to use `stdsimd`, as like all things, it'll eventually become stable
